letter-shop.md

```diff
@@ -0,0 +1,255 @@
+# Letter Shop Tutorial
+
+## What you need before you start:
+
+* a laptop with an internet connection and NodeJS installed
+* basic knowledge of the command line terminal
+* basic knowledge of JavaScript, including Promises and Buffers
+* (optional) the concept of one-way hash functions
+
+## What you'll learn:
+
+* the concept Hash Time Lock Agreements (HTLAs)
+* the Ledger Plugin Interface (LPI)
+* how to build a paid web app using Interledger!
+
+## Step 1: Creating the Letter Shop
+
+Save the following JavaScript as `shop.js`:
+
+```js
+const http = require('http')
+const crypto = require('crypto')
+const Plugin = require('ilp-plugin-xrp-escrow')
+function base64url (buf) { return buf.toString('base64').replace(/\+/g, '-').replace(/\//g, '_').replace(/=+$/, '') }
+
+let fulfillments = {}
+let letters = {}
+
+const plugin = new Plugin({
+  secret: 'ssGjGT4sz4rp2xahcDj87P71rTYXo',
+  account: 'rrhnXcox5bEmZfJCHzPxajUtwdt772zrCW',
+  server: 'wss://s.altnet.rippletest.net:51233',
+  prefix: 'test.crypto.xrp.'
+})
+
+plugin.connect().then(function () {
+  plugin.on('incoming_prepare', function (transfer) {
+    if (transfer.amount !== '10') {
+      plugin.rejectIncomingTransfer(transfer.id, {
+        code: 'F04',
+        name: 'Insufficient Destination Amount',
+        message: 'Please send exactly 10 drops, you sent ' + transfer.amount,
+        triggered_by: plugin.getAccount(),
+        triggered_at: new Date().toISOString(),
+        forwarded_by: [],
+        additional_info: {}
+      })
+    } else {
+      // the ledger will check if the fulfillment is correct and if it was submitted before the transfer's
+      // rollback timeout
+      plugin.fulfillCondition(transfer.id, fulfillments[transfer.executionCondition]).catch(function () {})
+    }
+  })
+
+  http.createServer(function (req, res) {
+    if (letters[req.url.substring(1)]) {
+      res.end('Your letter: ' + letters[req.url.substring(1)])
+    } else {
+      const secret = crypto.randomBytes(32)
+      const fulfillment = base64url(secret)
+      const condition = base64url(crypto.createHash('sha256').update(secret).digest())
+      const letter = ('ABCDEFGHIJKLMNOPQRSTUVWXYZ').split('')[(Math.floor(Math.random() * 26))]
+      fulfillments[condition] = fulfillment
+      letters[fulfillment] = letter
+      console.log('Generated letter for visitor on ', req.url, { secret, fulfillment, condition, letter })
+      res.end('Please send an Interledger payment to ' + plugin.getAccount() + ' with amount 10 drops and condition ' + condition)
+    }
+  }).listen(8000)
+})
+```
```

Comment:
I'm trying to read this from the perspective of someone who hasn't seen our code before. It feels like this code could use more explanation of what it's doing. What do you think about either:
1. Walking people through building up this example, explaining what each step is doing or
2. Breaking down the code segment by segment after you show the full example or
3. Adding more comments in the code about what's going on?

Reply:
yeah, if we tell people to clone the repo then we have more freedom to break down the snippets into smaller ones.